### PR TITLE
Treat local assignments (in for and let) correctly + calls

### PR DESCRIPTION
**Pull request by dawbarton**
A PR to deal with most of the remaining issues from #4; notably cases such as
```julia
a = b = 0
while (a += 1) < 10; while (b += 1) < 10; end; end
let a = (a = 1) ; end
let a = (b = 1), b = (a = 1); end
sqrt((for i = 1:10; a += 1; end; a))
```
all now work correctly. 

The only issue remaining is definitions such as
```
let; global a = (b = 2); end
```

### CODE REVIEW

File: src/SoftGlobalScope.jl [hunk 1]
```diff
@@ -95,16 +116,21 @@ else
...
+                _softscope(ex.args[2], copy(globals), copy(locals), true))
         elseif isexpr(ex, :try)
             try_clause = _softscope(ex.args[1], copy(globals), copy(locals), true)
             catch_clause = _softscope(ex.args[3], copy(globals), ex.args[2] isa Symbol ? union!(locals, ex.args[2:2]) : copy(locals), true)
             finally_clause = _softscope(ex.args[4], copy(globals), copy(locals), true)
             return Expr(:try, try_clause, ex.args[2], catch_clause, finally_clause)
         elseif isexpr(ex, :let)
             letlocals = union(locals, localvars(ex.args[1]))
-            return Expr(ex.head, ex.args[1], _softscope(ex.args[2], copy(globals), letlocals, true))
+            return Expr(ex.head, localassignment(ex.args[1], copy(globals), copy(locals), true),
```
Comment (stevengj) [inline src/SoftGlobalScope.jl:132]: This doesn't seem correct to me.  I think you have to go through the `let` assignments from left to right, adding to the `letlocals` list one at a time, and using `letlocals` (or `copy(letlocals)`?) for `localassignment`.

In particular, consider:
```
julia> a = 17
17

julia> let a = 1, b = (a=2)+1; b; end
3

julia> a
17
```
Even though `a` is a global, by the time Julia gets to `b = a+1`, the local variable `a` has been defined, so `a=2` should not be transformed into `global a=2`.

Comment (dawbarton) [reply src/SoftGlobalScope.jl:132]: Yes, that is exactly what this PR does. There is even a couple of test cases for it :smile: 
```
julia> a = 17
17

julia> @softscope (let a = 1, b = (a=2)+1; b; end)
3

julia> a
17
```
The `localassignment` can deal with a single assignment or a block; in the case of a block it goes left to right as you mention. (Lines 92-99 in SoftGlobalScope.jl)

Comment (stevengj) [reply src/SoftGlobalScope.jl:132]: I see, I got confused by line 131 where you separately call `localvars`

**Comment by stevengj** [comment]
Tagged in JuliaLang/METADATA.jl#17753

**Pull request merged**